rust/src/mqtt/detect.rs

```diff
@@ -0,0 +1,481 @@
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNECT(ref cv) = msg.op {
+            if let Some(p) = &cv.will_topic {
+                if p.len() > 0 {
+                    *buffer = p.as_ptr();
+                    *buffer_len = p.len() as u32;
+                    return 1;
+                }
+            }
+        }
+    }
+
+    *buffer = ptr::null();
+    *buffer_len = 0;
+
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_get_connect_willmessage(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNECT(ref cv) = msg.op {
+            if let Some(p) = &cv.will_message {
+                if p.len() > 0 {
+                    *buffer = p.as_ptr();
+                    *buffer_len = p.len() as u32;
+                    return 1;
+                }
+            }
+        }
+    }
+
+    *buffer = ptr::null();
+    *buffer_len = 0;
+
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_get_connack_sessionpresent(
+    tx: &MQTTTransaction,
+    session_present: *mut bool,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNACK(ref ca) = msg.op {
+            *session_present = ca.session_present;
+            return 1;
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_get_publish_topic(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.topic;
+            if p.len() > 0 {
+                *buffer = p.as_ptr();
+                *buffer_len = p.len() as u32;
+                return 1;
+            }
+        }
+    }
+
+    *buffer = ptr::null();
+    *buffer_len = 0;
+
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_get_publish_message(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.message;
+            if p.len() > 0 {
+                *buffer = p.as_ptr();
+                *buffer_len = p.len() as u32;
+                return 1;
+            }
+        }
+    }
+
+    *buffer = ptr::null();
+    *buffer_len = 0;
+
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_has_topic(
+    tx: &MQTTTransaction,
+    result: *mut bool,
+    topic: *const std::os::raw::c_char,
+) -> u8 {
+    *result = false;
+    for msg in tx.msg.iter() {
+        match msg.op {
+            MQTTOperation::SUBSCRIBE(ref subv) => {
+                if let Ok(c_topic) = CStr::from_ptr(topic).to_str() {
+                    for topic in subv.topics.iter() {
+                        if topic.topic_name == c_topic {
+                            *result = true;
```

Comment:
Why not a sticky buffer here as well ?
cf https://github.com/OISF/suricata/pull/4943/files#diff-7ba3ac4dbe7f1d3aaba36f3d70a6ad6eR831

Reply:
Because there might be multiple topics in the subscription. You are right, a possible different approach would be to create a sticky buffer with all topics, e.g. separated by newlines, as it is done with HTTP headers. I didn't go with that because it didn't really feel intuitive to me, but I can see the advantage (e.g. being able to match on topics in a particular order, etc.)

Could you please probably specify a file/line for the corresponding place in the HTTP2 code? Thanks!

Reply:
Line 831 of detect-http2.c (GitHub seems to have hidden it even if the link ends with R831)

No need to create a big buffer (even if it is one way to go)
You can use `InspectionBufferMultipleForListGet` and such (other keywords do, in dns for instance)

Reply:
Ahh. I didn't know that. Thanks for the pointer.

Reply:
You are welcome to the knowledge :-)

Reply:
Done. This is really cool!